### PR TITLE
netpbm: Fix cross platform build

### PR DESCRIPTION
**Pull request by kampka**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kampka** [comment]
@GrahamcOfBorg build netpbm
@GrahamcOfBorg build pkgsCross.aarch64-multiplatform.netpbm

**Comment by jtojnar** [comment]
I tried updating the package (#97045) if it makes any difference but unfortunately not will e-mail the maintainer.

File: pkgs/tools/graphics/netpbm/default.nix [hunk 1]
```diff
@@ -27,16 +28,43 @@ stdenv.mkDerivation {
...
+      substituteInPlace GNUmakefile \
+          --replace 'pkg-config' '${buildPackages.pkgconfig}/bin/${buildPackages.pkgconfig.targetPrefix}pkg-config'
+
+      substituteInPlace config.mk.in \
+          --replace 'AR = ar' 'AR = ${stdenv.lib.getBin stdenv.cc.bintools.bintools}/bin/${stdenv.cc.targetPrefix}ar'
+      substituteInPlace config.mk.in \
+          --replace 'RANLIB = ranlib' 'RANLIB = ${stdenv.lib.getBin stdenv.cc.bintools.bintools}/bin/${stdenv.cc.targetPrefix}ranlib'
+
+      substituteInPlace buildtools/install.sh \
+          --replace 'STRIPPROG-strip' 'STRIPPROG:-${stdenv.lib.getBin stdenv.cc.bintools.bintools}/bin/${stdenv.cc.targetPrefix}strip'
```
Comment (jtojnar) [inline pkgs/tools/graphics/netpbm/default.nix:52]: Looks like setting environment variable `STRIPPROG = lib.optionalString (stdenv.buildPlatform != stdenv.hostPlatform)  "${stdenv.lib.getBin stdenv.cc.bintools.bintools}/bin/${stdenv.cc.targetPrefix}strip";` should work.

File: pkgs/tools/graphics/netpbm/default.nix [hunk 2]
```diff
@@ -27,16 +28,43 @@ stdenv.mkDerivation {
...
+      substituteInPlace buildtools/configure.pl \
+          --replace 'pkg-config' '${buildPackages.pkgconfig}/bin/${buildPackages.pkgconfig.targetPrefix}pkg-config'
+
+      substituteInPlace GNUmakefile \
+          --replace 'pkg-config' '${buildPackages.pkgconfig}/bin/${buildPackages.pkgconfig.targetPrefix}pkg-config'
+
+      substituteInPlace config.mk.in \
+          --replace 'AR = ar' 'AR = ${stdenv.lib.getBin stdenv.cc.bintools.bintools}/bin/${stdenv.cc.targetPrefix}ar'
+      substituteInPlace config.mk.in \
+          --replace 'RANLIB = ranlib' 'RANLIB = ${stdenv.lib.getBin stdenv.cc.bintools.bintools}/bin/${stdenv.cc.targetPrefix}ranlib'
```
Comment (jtojnar) [inline pkgs/tools/graphics/netpbm/default.nix:49]: I would just append them to the generated `config.mk` like we do with the rest configure flags.

File: pkgs/tools/graphics/netpbm/default.nix [hunk 3]
```diff
@@ -27,16 +28,43 @@ stdenv.mkDerivation {
...
+
+      substituteInPlace buildtools/install.sh \
+          --replace 'STRIPPROG-strip' 'STRIPPROG:-${stdenv.lib.getBin stdenv.cc.bintools.bintools}/bin/${stdenv.cc.targetPrefix}strip'
+    '');
+
+  preBuild = ''
+    pushd buildtools
+    make CC=${buildPackages.stdenv.cc}/bin/${buildPackages.stdenv.cc.targetPrefix}cc
```
Comment (jtojnar) [inline pkgs/tools/graphics/netpbm/default.nix:57]: Could you clarify why is this needed? Would not the default `CC = cc` in `config.mk` already be  the same?

```
nix-repl> pkgsCross.aarch64-multiplatform.buildPackages.stdenv.cc.targetPrefix       
""
```

File: pkgs/tools/graphics/netpbm/default.nix [hunk 4]
```diff
@@ -27,16 +28,43 @@ stdenv.mkDerivation {
...
+    popd
   '';
 
+
   nativeBuildInputs = [
     pkgconfig
     flex
     makeWrapper
+    libpng  # for libpng-config
+    libxml2 # for xml2-config
```
Comment (jtojnar) [inline pkgs/tools/graphics/netpbm/default.nix:67]: Looking at the `Makefile`, I would expect the config-tools to be only used when pkg-config fails. Could you share the error message?

File: pkgs/tools/graphics/netpbm/default.nix (hunk 3 above)
Comment (jtojnar) [reply pkgs/tools/graphics/netpbm/default.nix:57]: Oh, looks like during cross-compilation there are only the prefixed binaries.

Though using `CC` environment variable might work.

Comment (kampka) [reply pkgs/tools/graphics/netpbm/default.nix:57]: During cross-platform build, the buildtools subpackage needs to be built with the local native compiler as those tools will be run during build, while the rest of the package will need to be built using the target platform compiler.
That is why I am building them separately like this.

File: pkgs/tools/graphics/netpbm/default.nix (hunk 1 above)
Comment (kampka) [reply pkgs/tools/graphics/netpbm/default.nix:52]: You are right, I assumed the same. It does not work though and I haven't investigated very deeply on why that is.

File: pkgs/tools/graphics/netpbm/default.nix (hunk 2 above)
Comment (kampka) [reply pkgs/tools/graphics/netpbm/default.nix:49]: The difference is that the rest of the configure flags are commented out in the `config.mk` while AR and RANLIB are not.
It might work to just append, but I am not very familiar with the semantics here, so I opted for the most explicit solution.

File: pkgs/tools/graphics/netpbm/default.nix (hunk 4 above)
Comment (kampka) [reply pkgs/tools/graphics/netpbm/default.nix]: You are right, for some reason, I can't reproduce it any longer.
Pushed an update to remove these.

**Comment by kampka** [comment]
@GrahamcOfBorg build netpbm
@GrahamcOfBorg build pkgsCross.aarch64-multiplatform.netpbm

File: pkgs/tools/graphics/netpbm/default.nix (hunk 2 above)
Comment (jtojnar) [reply pkgs/tools/graphics/netpbm/default.nix:49]: The resulting file just gets included into Makefile, so later definitions should override the previous ones.

File: pkgs/tools/graphics/netpbm/default.nix (hunk 3 above)
Comment (jtojnar) [reply pkgs/tools/graphics/netpbm/default.nix:57]: Right, I did not notice the directory change.

Actually, it looks like the build tools makefile uses `CC_FOR_BUILD` variable so setting it in `config.mk` should remove the necessity to invoke `make` manually.

Comment (kampka) [reply pkgs/tools/graphics/netpbm/default.nix:57]: You are right, it should work, but the build fails for buildtools with correct `CC_FOR_BUILD` and `LD_FOR_BUILD` set in `config.mk`. The workaround in preBuild works, though I haven't investigated any further why `config.mk` doesn't.

File: pkgs/tools/graphics/netpbm/default.nix [hunk 5]
```diff
@@ -27,12 +28,37 @@ stdenv.mkDerivation {
...
+  makeFlags = [
+    "CC=${stdenv.cc}/bin/${stdenv.cc.targetPrefix}cc"
+  ];
+
   postPatch = ''
     # Install libnetpbm.so symlink to correct destination
     substituteInPlace lib/Makefile \
       --replace '/sharedlink' '/lib'
+  '' + (stdenv.lib.optionalString (stdenv.buildPlatform != stdenv.hostPlatform) ''
+      substituteInPlace buildtools/configure.pl \
```
Comment (jtojnar) [inline pkgs/tools/graphics/netpbm/default.nix:40]: We do not use `buildtools/configure.pl` since it only supports interactive mode.

**Comment by jtojnar** [comment]
I tried the following patch (cross by default):

```patch
diff --git a/pkgs/tools/graphics/netpbm/default.nix b/pkgs/tools/graphics/netpbm/default.nix
index 4507c4fed85..67167c8193d 100644
--- a/pkgs/tools/graphics/netpbm/default.nix
+++ b/pkgs/tools/graphics/netpbm/default.nix
@@ -12,6 +12,7 @@
 , libtiff
 , enableX11 ? false
 , libX11
+, buildPackages
 }:
 
 stdenv.mkDerivation {
@@ -48,6 +49,10 @@ stdenv.mkDerivation {
     # Install libnetpbm.so symlink to correct destination
     substituteInPlace lib/Makefile \
       --replace '/sharedlink' '/lib'
+
+    # Respect the path set in config.mk
+    substituteInPlace GNUmakefile \
+      --replace 'pkg-config' '$(PKG_CONFIG)'
   '';
 
   configurePhase = ''
@@ -66,11 +71,19 @@ stdenv.mkDerivation {
 
     # Fix path to rgb.txt
     echo "RGB_DB_PATH = $out/share/netpbm/misc/rgb.txt" >> config.mk
+
+    # For cross-compilation
+    echo "CC = ${stdenv.cc}/bin/${stdenv.cc.targetPrefix}cc" >> config.mk
+    echo "AR = ${stdenv.lib.getBin stdenv.cc.bintools.bintools}/bin/${stdenv.cc.targetPrefix}ar" >> config.mk
+    echo "PKG_CONFIG = ${buildPackages.pkgconfig}/bin/${buildPackages.pkgconfig.targetPrefix}pkg-config" >> config.mk
+    echo "RANLIB = ${stdenv.lib.getBin stdenv.cc.bintools.bintools}/bin/${stdenv.cc.targetPrefix}ranlib" >> config.mk
+    echo "CC_FOR_BUILD = ${buildPackages.stdenv.cc}/bin/${buildPackages.stdenv.cc.targetPrefix}cc" >> config.mk
+    echo "LD_FOR_BUILD = ${buildPackages.binutils}/bin/${buildPackages.binutils.targetPrefix}ld" >> config.mk
   '' + stdenv.lib.optionalString stdenv.isDarwin ''
     echo "LDSHLIB=-dynamiclib -install_name $out/lib/libnetpbm.\$(MAJ).dylib" >> config.mk
     echo "NETPBMLIBTYPE = dylib" >> config.mk
     echo "NETPBMLIBSUFFIX = dylib" >> config.mk
-
+  '' + ''
     runHook postConfigure
   '';
 
@@ -94,6 +107,9 @@ stdenv.mkDerivation {
     runHook postInstall
   '';
 
+  # Environment variables
+  STRIPPROG = "${stdenv.lib.getBin stdenv.cc.bintools.bintools}/bin/${stdenv.cc.targetPrefix}strip";
+
   passthru.updateScript = ./update.sh;
 
   meta = {
```

But it fails for `native`:

```
/nix/store/jgcv40jcr1v9bmnj070mgvyi6m7r7w5r-gcc-wrapper-9.3.0/bin/cc -c -o typegen.o  typegen.c
/nix/store/r6wg5hrz4kxcdjmjsbxn9xr55zxzpr07-binutils-wrapper-2.31.1/bin/ld -o typegen  typegen.o
/nix/store/zp4vhfn31ky68xy0s6mssxh4c90z9v9a-binutils-2.31.1/bin/ld: warning: cannot find entry symbol _start; defaulting to 0000000000401000
/nix/store/zp4vhfn31ky68xy0s6mssxh4c90z9v9a-binutils-2.31.1/bin/ld: typegen.o: in function `main':
typegen.c:(.text.startup+0xc): undefined reference to `puts'
/nix/store/zp4vhfn31ky68xy0s6mssxh4c90z9v9a-binutils-2.31.1/bin/ld: typegen.c:(.text.startup+0x18): undefined reference to `puts'
/nix/store/zp4vhfn31ky68xy0s6mssxh4c90z9v9a-binutils-2.31.1/bin/ld: typegen.c:(.text.startup+0x24): undefined reference to `puts'
/nix/store/zp4vhfn31ky68xy0s6mssxh4c90z9v9a-binutils-2.31.1/bin/ld: typegen.c:(.text.startup+0x30): undefined reference to `puts'
/nix/store/zp4vhfn31ky68xy0s6mssxh4c90z9v9a-binutils-2.31.1/bin/ld: typegen.c:(.text.startup+0x3c): undefined reference to `puts'
/nix/store/zp4vhfn31ky68xy0s6mssxh4c90z9v9a-binutils-2.31.1/bin/ld: typegen.o:typegen.c:(.text.startup+0x48): more undefined references to `puts' follow
make[1]: *** [/build/advanced-r3735/buildtools/Makefile:47: typegen] Error 1
```

and for `cross`:

```
/nix/store/jgcv40jcr1v9bmnj070mgvyi6m7r7w5r-gcc-wrapper-9.3.0/bin/cc -c -o typegen.o  typegen.c
/nix/store/0wry6m9bb5gm07s71pvf8l0cmfpay3gw-aarch64-unknown-linux-gnu-binutils-wrapper-2.31.1/bin/aarch64-unknown-linux-gnu-ld -o typegen  typegen.o
/nix/store/c178lyxkahsr67s48wc2r4ynii5q05qw-aarch64-unknown-linux-gnu-binutils-2.31.1/bin/aarch64-unknown-linux-gnu-ld: typegen.o: Relocations in generic ELF (EM: 62)
/nix/store/c178lyxkahsr67s48wc2r4ynii5q05qw-aarch64-unknown-linux-gnu-binutils-2.31.1/bin/aarch64-unknown-linux-gnu-ld: typegen.o: Relocations in generic ELF (EM: 62)
/nix/store/c178lyxkahsr67s48wc2r4ynii5q05qw-aarch64-unknown-linux-gnu-binutils-2.31.1/bin/aarch64-unknown-linux-gnu-ld: typegen.o: error adding symbols: file in wrong format
make[1]: *** [/build/advanced-r3735/buildtools/Makefile:47: typegen] Error 1
```

**Comment by kampka** [comment]
@jtojnar That is pretty much the same results I am getting.
So how do you want to procede with this?

**Comment by kampka** [comment]
@GrahamcOfBorg build netpbm
@GrahamcOfBorg build pkgsCross.aarch64-multiplatform.netpbm

**Comment by jtojnar** [comment]
Managed to get it building mostly using config.mk in #97045

**Comment by kampka** [comment]
Closing this in favor of #97045
